packages/message-manager/src/EncryptionPublicKeyManager.ts

```diff
@@ -0,0 +1,149 @@
+
+/**
+ * @type MessageParamsMetamask
+ *
+ * Represents the parameters to pass to the eth_getEncryptionPublicKey method once the request is approved
+ * plus data added by MetaMask.
+ * @property metamaskId - Added for tracking and identification within MetaMask
+ * @property data - Encryption public key
+ * @property from - Address from which to extract the encryption public key
+ * @property origin? - Added for request origin identification
+ */
+export interface EncryptionPublicKeyParamsMetamask
+  extends AbstractMessageParamsMetamask {
+  data: string;
+}
+
+/**
+ * Controller in charge of managing - storing, adding, removing, updating - Messages.
+ */
+export class EncryptionPublicKeyManager extends AbstractMessageManager<
+  EncryptionPublicKey,
+  EncryptionPublicKeyParams,
+  EncryptionPublicKeyParamsMetamask
+> {
+  /**
+   * Name of this controller used during composition
+   */
+  override name = 'EncryptionPublicKeyManager';
+
+  /**
+   * Creates a new Message with an 'unapproved' status using the passed messageParams.
+   * this.addMessage is called to add the new Message to this.messages, and to save the unapproved Messages.
+   *
+   * @param messageParams - The params for the eth_getEncryptionPublicKey call to be made after the message is approved.
+   * @param req - The original request object possibly containing the origin.
+   * @returns Promise resolving to the raw data of the request.
+   */
+  addUnapprovedMessageAsync(
+    messageParams: EncryptionPublicKeyParams,
+    req?: OriginalRequest,
+  ): Promise<string> {
+    return new Promise((resolve, reject) => {
+      validateEncryptionPublicKeyMessageData(messageParams);
+      const messageId = this.addUnapprovedMessage(messageParams, req);
+      this.hub.once(`${messageId}:finished`, (data: EncryptionPublicKey) => {
+        switch (data.status) {
+          case 'received':
+            return resolve(data.rawSig as string);
+          case 'rejected':
+            return reject(
+              new Error(
+                'MetaMask EncryptionPublicKey: User denied message EncryptionPublicKey.',
+              ),
+            );
+          default:
+            return reject(
+              new Error(
+                `MetaMask EncryptionPublicKey: Unknown problem: ${JSON.stringify(
+                  messageParams,
+                )}`,
+              ),
+            );
+        }
+      });
+    });
+  }
+
+  /**
+   * Creates a new Message with an 'unapproved' status using the passed messageParams.
+   * this.addMessage is called to add the new Message to this.messages, and to save the
+   * unapproved Messages.
+   *
+   * @param messageParams - The params for the eth_getEncryptionPublicKey call to be made after the message
+   * is approved.
+   * @param req - The original request object possibly containing the origin.
+   * @returns The id of the newly created message.
+   */
+  addUnapprovedMessage(
```

Comment:
Just as a heads up, depending on which is merged first, this is now `async` following the signature provider request support.

Reply:
Won't it look a bit odd if we have a `addUnapprovedMessageAsync` and a `addUnapprovedMessage` that are both async?

Reply:
The former doesn't mean async in the promise sense, but that it won't resolve until the message status is considered finished. Not ideal, we could rename with a breaking change down the line.